File: usage/benchmarks/trajectory_storage/codecs.py
"""Lossless benchmark codecs with common metadata and typed variable events.

The fixed-width candidate uses a protobuf variable-event sidecar, avoiding a
second handwritten nested-event schema. Binary layouts are little endian.
"""
from __future__ import annotations

import json
import math
import struct
from dataclasses import dataclass
from pathlib import Path

from google.protobuf import descriptor_pb2, descriptor_pool, message_factory

COMMON = (
    ("step", "sint64", 0), ("action", "sint32", 0), ("reward", "double", 0.0),
    ("done", "bool", False), ("cursor_x", "uint32", 0),
    ("cursor_y", "uint32", 0), ("cursor_z", "uint32", 0),
    ("voxel_count", "uint64", 0), ("placed", "bool", False),
)
MUTATION = (("occupied", "bool", False), ("kind", "uint32", 0),
            ("active", "bool", False), ("reward_weight", "double", 0.0))
VARIABLE = ("mutations", "actions", "rewards", "cursors", "placed_per_agent", "dones")
RECORD = struct.Struct("<Hqid?IIIQ?")
OFFSET = struct.Struct("<Q")
HEADER = struct.Struct("<4sIQ")
MAGIC = b"TSB1"


def json_bytes(value, *, pretty=False):
    return json.dumps(value, allow_nan=False, ensure_ascii=True,
                      indent=2 if pretty else None,
                      separators=None if pretty else (",", ":")).encode("utf-8")


def _messages():
    """Build the checked-in schema without requiring protoc at runtime."""
    file = descriptor_pb2.FileDescriptorProto(name="trajectory_bench.proto",
                                             package="trajectory_bench", syntax="proto3")
    types = descriptor_pb2.FieldDescriptorProto
    def message(name):
        return file.message_type.add(name=name)
    def field(msg, name, number, kind, repeated=False, target=None):
        item = msg.field.add(name=name, number=number,
                             type=getattr(types, "TYPE_" + kind.upper()),
                             label=types.LABEL_REPEATED if repeated else types.LABEL_OPTIONAL)
        if target:
            item.type_name = ".trajectory_bench." + target
    coord = message("Coord")
    field(coord, "values", 1, "uint32", True)
    mutation = message("Mutation")
    for i, (name, kind, _) in enumerate(MUTATION, 1):
        field(mutation, name, i, kind)
    field(mutation, "coordinate", 5, "uint32", True)
    field(mutation, "presence", 15, "uint32")
    field(mutation, "extras_json", 16, "bytes")
    step = message("Step")
    for i, (name, kind, _) in enumerate(COMMON, 1):
        field(step, name, i, kind)
    field(step, "presence", 15, "uint32")
    field(step, "extras_json", 16, "bytes")
    field(step, "mutations", 20, "message", True, "Mutation")
    field(step, "actions", 21, "sint32", True)
    field(step, "rewards", 22, "double", True)
    field(step, "cursors", 23, "message", True, "Coord")
    field(step, "placed_per_agent", 24, "bool", True)
    field(step, "dones", 25, "bool", True)
    trajectory = message("Trajectory")
    field(trajectory, "metadata_json", 1, "bytes")
    field(trajectory, "steps", 2, "message", True, "Step")
    field(trajectory, "version", 3, "uint32")
    pool = descriptor_pool.DescriptorPool()
    pool.Add(file)
    return tuple(message_factory.GetMessageClass(pool.FindMessageTypeByName(
        "trajectory_bench." + name)) for name in ("Step", "Trajectory"))


StepMessage, TrajectoryMessage = _messages()


def encode_step(value):
    doubles = [value.get("reward", 0.0), *value.get("rewards", [])]
    doubles.extend(event.get("reward_weight", 0.0) for event in value.get("mutations", []))
    if not all(math.isfinite(number) for number in doubles):
        raise ValueError("nonfinite trajectory numeric field")
    msg = StepMessage()
    for i, (name, _, _) in enumerate(COMMON):
        if name in value:
            setattr(msg, name, value[name])
            msg.presence |= 1 << i
    for i, name in enumerate(VARIABLE, len(COMMON)):
        if name not in value:
            continue
        msg.presence |= 1 << i
        if name == "mutations":
            for event in value[name]:
                item = msg.mutations.add()
                for j, (key, _, _) in enumerate(MUTATION):
                    if key in event:
                        setattr(item, key, event[key])
                        item.presence |= 1 << j
                if "coordinate" in event:
                    item.coordinate.extend(event["coordinate"])
                    item.presence |= 1 << len(MUTATION)
                extras = {k: v for k, v in event.items()
                          if k not in {x[0] for x in MUTATION} | {"coordinate"}}
                if extras:
                    item.extras_json = json_bytes(extras)
        elif name == "cursors":
            for cursor in value[name]:
                msg.cursors.add().values.extend(cursor)
        else:
            getattr(msg, name).extend(value[name])
    extras = {k: v for k, v in value.items()
              if k not in {x[0] for x in COMMON} | set(VARIABLE)}
    if extras:
        msg.extras_json = json_bytes(extras)
    return msg


def decode_step(msg):
    value = json.loads(msg.extras_json) if msg.extras_json else {}
    for i, (name, _, _) in enumerate(COMMON):
        if msg.presence & (1 << i):
            value[name] = getattr(msg, name)
    for i, name in enumerate(VARIABLE, len(COMMON)):
        if not msg.presence & (1 << i):
            continue
        if name == "mutations":
            events = []
            for item in msg.mutations:
                event = json.loads(item.extras_json) if item.extras_json else {}
                for j, (key, _, _) in enumerate(MUTATION):
                    if item.presence & (1 << j):
                        event[key] = getattr(item, key)
                if item.presence & (1 << len(MUTATION)):
                    event["coordinate"] = list(item.coordinate)
                events.append(event)
            value[name] = events
        elif name == "cursors":
            value[name] = [list(cursor.values) for cursor in msg.cursors]
        else:
            value[name] = list(getattr(msg, name))
    return value


def split(payload):
    metadata = dict(payload)
    metadata["episode"] = dict(payload["episode"])
    steps = metadata["episode"].pop("steps")
    return metadata, steps


def join(metadata, steps):
    result = dict(metadata)
    result["episode"] = dict(metadata["episode"], steps=steps)
    return result


@dataclass
class Loaded:
    metadata: dict
    steps: object
    kind: str
    events: bytes = b""
    offsets: tuple = ()

    def __len__(self):
        return len(self.steps) // RECORD.size if self.kind == "binary" else len(self.steps)

    def step(self, index):
        if not 0 <= index < len(self):
            raise IndexError(index)
        if self.kind == "json":
            return self.steps[index]
        if self.kind == "protobuf":
            return decode_step(self.steps[index])
        values = RECORD.unpack_from(self.steps, index * RECORD.size)
        mask = values[0]
        msg = StepMessage.FromString(self.events[self.offsets[index]:self.offsets[index + 1]])
        result = decode_step(msg)
        for i, (name, _, _) in enumerate(COMMON):
            if mask & (1 << i):
                result[name] = values[i + 1]
        return result

    def export(self):
        return join(self.metadata, [self.step(i) for i in range(len(self))])

    def overlay(self, index):
        """Same cumulative coordinate/occupancy resolution as Rust replay."""
        if not 0 <= index < len(self):
            raise IndexError(index)
        result = {}
        for i in range(index + 1):
            for mutation in self.step(i).get("mutations", []):
                result[tuple(mutation["coordinate"])] = mutation["occupied"]
        return result


FORMATS = ("json-pretty", "json-zstd", "protobuf", "protobuf-zstd", "binary", "binary-zstd")


def _compress(value):
    import zstandard
    return zstandard.ZstdCompressor(level=3).compress(value)


def _decompress(value):
    import zstandard
    return zstandard.ZstdDecompressor().decompress(value)


def write(directory: Path, kind: str, payload: dict):
    if kind not in FORMATS:
        raise ValueError(kind)
    directory.mkdir(parents=True, exist_ok=True)
    metadata, steps = split(payload)
    if kind.startswith("json"):
        data = json_bytes(payload, pretty=kind == "json-pretty")
        (directory / "trajectory.data").write_bytes(_compress(data) if kind.endswith("zstd") else data)
        return
    if kind.startswith("protobuf"):
        msg = TrajectoryMessage(metadata_json=json_bytes(metadata), version=1)
        for step in steps:
            msg.steps.add().CopyFrom(encode_step(step))
        data = msg.SerializeToString(deterministic=True)
        (directory / "trajectory.data").write_bytes(_compress(data) if kind.endswith("zstd") else data)
        return
    records = bytearray(HEADER.pack(MAGIC, 1, len(steps)))
    events = bytearray()
    offsets = [0]
    for step in steps:
        if not math.isfinite(step.get("reward", 0.0)):
            raise ValueError("nonfinite trajectory reward")
        mask = sum(1 << i for i, (name, _, _) in enumerate(COMMON) if name in step)
        records.extend(RECORD.pack(mask, *(step.get(name, default) for name, _, default in COMMON)))
        event = {k: v for k, v in step.items() if k not in {x[0] for x in COMMON}}
        events.extend(encode_step(event).SerializeToString(deterministic=True))
        offsets.append(len(events))
    manifest = {"benchmark_version": 1, "metadata": metadata,
                "record_bytes": RECORD.size, "step_count": len(steps),
                "compression": "zstd-3" if kind.endswith("zstd") else "none"}
    (directory / "manifest.json").write_bytes(json_bytes(manifest, pretty=True))
    for name, data in (("steps.bin", records), ("events.pb", events),
                       ("offsets.bin", b"".join(OFFSET.pack(x) for x in offsets))):
        (directory / name).write_bytes(_compress(data) if kind.endswith("zstd") else data)


def read(directory: Path, kind: str):
    if kind not in FORMATS:
        raise ValueError(kind)
    def data(name):
        value = (directory / name).read_bytes()
        return _decompress(value) if kind.endswith("zstd") else value
    if kind.startswith("json"):
        metadata, steps = split(json.loads(data("trajectory.data")))
        return Loaded(metadata, steps, "json")
    if kind.startswith("protobuf"):
        msg = TrajectoryMessage.FromString(data("trajectory.data"))
        if msg.version != 1:
            raise ValueError("unsupported protobuf benchmark version")
        return Loaded(json.loads(msg.metadata_json), msg.steps, "protobuf")
    manifest = json.loads((directory / "manifest.json").read_bytes())
    records, events, index = data("steps.bin"), data("events.pb"), data("offsets.bin")
    magic, version, count = HEADER.unpack_from(records)
    if (magic != MAGIC or version != 1 or manifest["benchmark_version"] != 1
            or manifest["record_bytes"] != RECORD.size or manifest["step_count"] != count):
        raise ValueError("invalid binary benchmark header")
    if len(records) != HEADER.size + count * RECORD.size or len(index) != (count + 1) * OFFSET.size:
        raise ValueError("truncated binary records/index")
    offsets = tuple(x[0] for x in OFFSET.iter_unpack(index))
    if offsets[0] != 0 or offsets[-1] != len(events) or any(a > b for a, b in zip(offsets, offsets[1:])):
        raise ValueError("invalid event offsets")
    return Loaded(manifest["metadata"], records[HEADER.size:], "binary", events, offsets)

File: usage/benchmarks/trajectory_storage/rust/src/codecs.rs
use std::{fs, path::Path};
use protobuf::Message;
use serde_json::{json, Map, Value};
mod generated { include!(concat!(env!("OUT_DIR"), "/protos/mod.rs")); }
use generated::trajectory_bench::{Mutation, Step, Trajectory};

type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;

fn extras(bytes: &[u8]) -> Result<Map<String, Value>> {
    if bytes.is_empty() { Ok(Map::new()) } else { Ok(serde_json::from_slice(bytes)?) }
}

fn mutation(item: &Mutation) -> Result<Value> {
    let mut out = extras(&item.extras_json)?;
    for (i, (key, value)) in [
        ("occupied", json!(item.occupied)), ("kind", json!(item.kind)),
        ("active", json!(item.active)), ("reward_weight", json!(item.reward_weight)),
        ("coordinate", json!(item.coordinate)),
    ].into_iter().enumerate() {
        if item.presence & (1 << i) != 0 { out.insert(key.into(), value); }
    }
    Ok(Value::Object(out))
}

fn step(item: &Step) -> Result<Value> {
    let mut out = extras(&item.extras_json)?;
    for (i, (key, value)) in [
        ("step", json!(item.step)), ("action", json!(item.action)),
        ("reward", json!(item.reward)), ("done", json!(item.done)),
        ("cursor_x", json!(item.cursor_x)), ("cursor_y", json!(item.cursor_y)),
        ("cursor_z", json!(item.cursor_z)), ("voxel_count", json!(item.voxel_count)),
        ("placed", json!(item.placed)),
    ].into_iter().enumerate() {
        if item.presence & (1 << i) != 0 { out.insert(key.into(), value); }
    }
    let variables = [
        ("mutations", Value::Array(item.mutations.iter().map(mutation).collect::<Result<_>>()?)),
        ("actions", json!(item.actions)), ("rewards", json!(item.rewards)),
        ("cursors", json!(item.cursors.iter().map(|c| &c.values).collect::<Vec<_>>())),
        ("placed_per_agent", json!(item.placed_per_agent)), ("dones", json!(item.dones)),
    ];
    for (i, (key, value)) in variables.into_iter().enumerate() {
        if item.presence & (1 << (i + 9)) != 0 { out.insert(key.into(), value); }
    }
    Ok(Value::Object(out))
}

fn bytes(path: &Path, compressed: bool) -> Result<Vec<u8>> {
    let raw = fs::read(path)?;
    Ok(if compressed { zstd::stream::decode_all(&raw[..])? } else { raw })
}

fn fixed_step(raw: &[u8]) -> Result<Value> {
    if raw.len() != 44 { return Err("invalid record length".into()); }
    let mask = u16::from_le_bytes(raw[0..2].try_into()?);
    let fields = [
        ("step", json!(i64::from_le_bytes(raw[2..10].try_into()?))),
        ("action", json!(i32::from_le_bytes(raw[10..14].try_into()?))),
        ("reward", json!(f64::from_le_bytes(raw[14..22].try_into()?))),
        ("done", json!(raw[22] != 0)),
        ("cursor_x", json!(u32::from_le_bytes(raw[23..27].try_into()?))),
        ("cursor_y", json!(u32::from_le_bytes(raw[27..31].try_into()?))),
        ("cursor_z", json!(u32::from_le_bytes(raw[31..35].try_into()?))),
        ("voxel_count", json!(u64::from_le_bytes(raw[35..43].try_into()?))),
        ("placed", json!(raw[43] != 0)),
    ];
    let mut out = Map::new();
    for (i, (key, value)) in fields.into_iter().enumerate() {
        if mask & (1 << i) != 0 { out.insert(key.into(), value); }
    }
    Ok(Value::Object(out))
}

pub fn read(dir: &Path, format: &str) -> Result<Value> {
    let compressed = format.ends_with("zstd");
    if format.starts_with("json") {
        return Ok(serde_json::from_slice(&bytes(&dir.join("trajectory.data"), compressed)?)?);
    }
    if format.starts_with("protobuf") {
        let msg = Trajectory::parse_from_bytes(&bytes(&dir.join("trajectory.data"), compressed)?)?;
        if msg.version != 1 { return Err("unsupported protobuf version".into()); }
        let mut payload: Value = serde_json::from_slice(&msg.metadata_json)?;
        payload["episode"]["steps"] = Value::Array(msg.steps.iter().map(step).collect::<Result<_>>()?);
        return Ok(payload);
    }
    if format != "binary" && format != "binary-zstd" { return Err("unknown format".into()); }
    let manifest: Value = serde_json::from_slice(&fs::read(dir.join("manifest.json"))?)?;
    let records = bytes(&dir.join("steps.bin"), compressed)?;
    let offsets = bytes(&dir.join("offsets.bin"), compressed)?;
    let events = bytes(&dir.join("events.pb"), compressed)?;
    if records.len() < 16 || &records[..4] != b"TSB1" || records[4..8] != 1u32.to_le_bytes()
        || manifest["benchmark_version"] != 1 || manifest["record_bytes"] != 44 {
        return Err("invalid binary header".into());
    }
    let count = usize::try_from(u64::from_le_bytes(records[8..16].try_into()?))?;
    if records.len().checked_sub(16) != count.checked_mul(44)
        || offsets.len() != count.checked_add(1).and_then(|n| n.checked_mul(8)).ok_or("index overflow")?
        || manifest["step_count"].as_u64() != Some(count as u64) {
        return Err("invalid binary lengths".into());
    }
    let index = offsets.chunks_exact(8).map(|b| usize::try_from(u64::from_le_bytes(b.try_into().unwrap())))
        .collect::<std::result::Result<Vec<_>, _>>()?;
    if index.first() != Some(&0) || index.last() != Some(&events.len()) || index.windows(2).any(|p| p[0] > p[1]) {
        return Err("invalid event index".into());
    }
    let mut steps = Vec::with_capacity(count);
    for (i, record) in records[16..].chunks_exact(44).enumerate() {
        let mut value = step(&Step::parse_from_bytes(&events[index[i]..index[i+1]])?)?;
        value.as_object_mut().ok_or("invalid event")?.extend(fixed_step(record)?.as_object().unwrap().clone());
        steps.push(value);
    }
    let mut payload = manifest["metadata"].clone();
    payload["episode"]["steps"] = Value::Array(steps);
    Ok(payload)
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn record_width_and_large_coordinates() {
        let mut raw = [0u8; 44];
        raw[..2].copy_from_slice(&(1u16 << 4).to_le_bytes());
        raw[23..27].copy_from_slice(&70_001u32.to_le_bytes());
        assert_eq!(fixed_step(&raw).unwrap(), json!({"cursor_x": 70001}));
        assert!(fixed_step(&raw[..43]).is_err());
    }
    #[test]
    fn protobuf_presence_and_events() {
        let mut item = Step::new();
        item.presence = (1 << 1) | (1 << 9);
        assert_eq!(step(&item).unwrap(), json!({"action":0,"mutations":[]}));
    }
}
